sqlstats/sqlstats_test.go

```diff
@@ -119,3 +121,62 @@ func TestCollect(t *testing.T) {
 		}
 	}
 }
+
+func TestTracerResetRace(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	tracer := &Tracer{}
+	db := sql.OpenDB(sqlite.Connector("file:"+t.TempDir()+"/test.db", nil, tracer))
+	defer db.Close()
+
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tx.Rollback()
+	if _, err := tx.ExecContext(ctx, "CREATE TABLE t (c);"); err != nil {
+		t.Fatal(err)
+	}
+	if err := tx.Commit(); err != nil {
+		t.Fatal(err)
+	}
+
+	// Continually reset the tracer.
+	go func() {
+		for ctx.Err() == nil {
+			tracer.Reset()
+		}
+	}()
+
+	// Continually grab transactions and insert into the database.
+	errc := make(chan error, 1)
+	go func() {
+		defer close(errc)
+		var n int64
+		for ctx.Err() == nil {
+			n++
+			tx, err = db.BeginTx(ctx, nil)
+			if err != nil {
+				errc <- fmt.Errorf("starting tx: %w", err)
+				return
+			}
+			if _, err = tx.Exec("INSERT INTO t VALUES(?)", n); err != nil {
+				errc <- fmt.Errorf("insert: %w", err)
+				return
+			}
+			if n%2 == 0 {
+				if err := tx.Commit(); err != nil {
+					errc <- fmt.Errorf("commit: %w", err)
+					return
+				}
+			}
+			tx.Rollback()
+		}
+	}()
+
+	err = <-errc
```

Comment:
Is there some reason not to loop over the channel? It is possible multiple errors could be reported, and we still want all the goroutines to be able to exit.

Reply:
There's only one goroutine writing to it now. Originally I was thinking that'd want to receive an error and ensure it's nil, but I didn't actually implement that behaviour - the channel is just closed with no sent errors when everything works successfully. I've switched to a loop.

Reply:
I was originally thinking of the panic handler which could write a second value, but now that's done. 😀